File: backend-js/src/settings/syncPlanning/syncService.ts
import { createHash, randomUUID } from "node:crypto";
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import type { MongoDBEventStore } from "@event-driven-io/emmett-mongodb";
import { toStreamName } from "@event-driven-io/emmett-mongodb";
import { stripReadModelListMetadata, stripReadModelMetadata } from "@/platform/readModelUtils.js";
import type { EntityCommandBus } from "@/es-kit/bus/EntityCommandBus.js";
import type { CreateTask, UpdateTask } from "@/task/application/commands.js";
import { Task } from "@/task/domain/Task.js";
import {
  TASK_LIST_PROJECTION_NAME,
  type TaskListDocument,
} from "@/task/projections/taskListProjection.js";
import { PROJECT_LIST_PROJECTION_NAME } from "@/project/projections/projectListProjection.js";
import { Project } from "@/project/domain/Project.js";
import type { UpsertSetting } from "@/settings/application/commands.js";
import {
  SETTINGS_PROJECTION_NAME,
  type SettingsDocument,
} from "@/settings/projections/settingsProjection.js";
import { Settings } from "@/settings/domain/Settings.js";
import { getKimiApiKey } from "@/ai/keyResolution.js";
import {
  DEFAULT_PLANNING_DIR,
  PLANNING_FOLDER_SETTING_KEY,
  PLANNING_SYNC_ENABLED_KEY,
  SYNC_HASH_SETTING_KEY,
} from "./constants.js";
import {
  buildProjectMap,
  guessProjectId,
  guessTaskGrouping,
} from "./guessTaskGrouping.js";
import { parseWeeklyPlan } from "./parseWeeklyPlan.js";
import type { ParsedWeeklyTask } from "./parseWeeklyPlan.js";

export type SyncWeeklyPlanResult = {
  status: "success" | "skipped";
  reason?: string;
  file_parsed: string;
  parser_mode: "regex" | "ai" | "skipped";
  tasks_created: number;
  tasks_updated: number;
  tasks_archived: number;
  total_parsed: number;
};

type StoredHashData = {
  file_name?: string;
  hash?: string;
};

export class SyncPlanningNotFoundError extends Error {
  readonly status = 404;

  constructor(message: string) {
    super(message);
    this.name = "SyncPlanningNotFoundError";
  }
}

/**
 * Loads the flat settings map from the inline projection.
 */
export async function loadSettingsMap(
  eventStore: MongoDBEventStore,
): Promise<SettingsDocument> {
  const streamName = toStreamName(Settings.streamType, Settings.GLOBAL_ID);
  const doc = await eventStore.projections.inline.findOne<SettingsDocument>({
    streamName,
    projectionName: SETTINGS_PROJECTION_NAME,
  });

  return doc ? stripReadModelMetadata(doc) : {};
}

/**
 * Upserts a single setting key via the bus.
 */
export async function upsertSettingValue(
  bus: EntityCommandBus,
  key: string,
  value: string,
  now: Date = new Date(),
): Promise<void> {
  const command: UpsertSetting = {
    type: "UpsertSetting",
    data: { key, value },
    metadata: { now },
  };

  await bus.send(command);
}

function resolvePlanningDir(settings: SettingsDocument): string {
  return settings[PLANNING_FOLDER_SETTING_KEY] || DEFAULT_PLANNING_DIR;
}

/**
 * Finds the latest `Current_Planning_*.md` file in the planning directory.
 */
export function getActivePlanningFile(planningDir: string): { filePath: string; fileName: string } {
  if (!existsSync(planningDir)) {
    throw new SyncPlanningNotFoundError(
      `Planning directory not found at: ${planningDir}`,
    );
  }

  const files = readdirSync(planningDir);
  const currentPlanningFiles = files.filter(
    (f) => f.startsWith("Current_Planning_") && f.endsWith(".md"),
  );

  if (currentPlanningFiles.length === 0) {
    throw new SyncPlanningNotFoundError(
      "No active weekly planning file (prefixed with 'Current_Planning_') found.",
    );
  }

  currentPlanningFiles.sort();
  const fileName = currentPlanningFiles[currentPlanningFiles.length - 1];
  return { filePath: join(planningDir, fileName), fileName };
}

function computeFileHash(content: string): string {
  return createHash("sha256").update(content, "utf-8").digest("hex");
}

function parseStoredHash(settings: SettingsDocument): StoredHashData {
  const raw = settings[SYNC_HASH_SETTING_KEY];
  if (!raw) {
    return {};
  }

  try {
    return JSON.parse(raw) as StoredHashData;
  } catch {
    return {};
  }
}

async function findTaskByNameAndSource(
  eventStore: MongoDBEventStore,
  name: string,
  source: string,
): Promise<TaskListDocument | null> {
  const tasks = stripReadModelListMetadata(
    await eventStore.projections.inline.find<TaskListDocument>(
      { streamType: Task.streamType, projectionName: TASK_LIST_PROJECTION_NAME },
      { name, source },
    ),
  );

  return tasks[0] ?? null;
}

async function findDonePlanningTasks(
  eventStore: MongoDBEventStore,
): Promise<TaskListDocument[]> {
  return stripReadModelListMetadata(
    await eventStore.projections.inline.find<TaskListDocument>(
      { streamType: Task.streamType, projectionName: TASK_LIST_PROJECTION_NAME },
      { source: "planning", status: "done", archived: false },
    ),
  );
}

async function loadProjectMap(eventStore: MongoDBEventStore): Promise<Map<string, string>> {
  const projects = stripReadModelListMetadata(
    await eventStore.projections.inline.find<{ name: string; id: string }>(
      { streamType: Project.streamType, projectionName: PROJECT_LIST_PROJECTION_NAME },
    ),
  );
  return buildProjectMap(projects);
}

async function handleCreateTask(
  bus: EntityCommandBus,
  taskData: ParsedWeeklyTask,
  projectId: string | null,
  now: Date,
): Promise<void> {
  const taskId = randomUUID();
  const grouping = guessTaskGrouping(
    taskData.name,
    taskData.description,
    taskData.category,
  );

  const command: CreateTask = {
    type: "CreateTask",
    data: {
      id: taskId,
      name: taskData.name,
      description: taskData.description,
      status: taskData.status,
      category: taskData.category,
      owner: taskData.owner,
      task_grouping: grouping,
      estimated_duration: 60,
      current_duration: 0,
      source: "planning",
      project_id: projectId,
      archived: taskData.status === "done",
    },
    metadata: { now },
  };

  // bus.send() triggers the handler + integration middleware (EAV upsert, etc.)
  await bus.send(command);
}

async function handleUpdateTask(
  bus: EntityCommandBus,
  existingTask: TaskListDocument,
  taskData: ParsedWeeklyTask,
  projectId: string | null,
  now: Date,
): Promise<void> {
  const grouping = guessTaskGrouping(
    taskData.name,
    taskData.description,
    taskData.category,
  );

  const patch: Record<string, unknown> = {
    description: taskData.description,
    status: taskData.status,
    category: taskData.category,
    owner: taskData.owner,
    task_grouping: grouping,
  };

  if (projectId) {
    patch.project_id = projectId;
  }

  if (taskData.status === "done") {
    patch.archived = true;
  }

  const command: UpdateTask = {
    type: "UpdateTask",
    data: { id: existingTask.id, patch },
    metadata: { now },
  };

  await bus.send(command);
}

async function handleArchiveTask(
  bus: EntityCommandBus,
  task: TaskListDocument,
  now: Date,
): Promise<void> {
  const patch = { archived: true };
  const command: UpdateTask = {
    type: "UpdateTask",
    data: { id: task.id, patch },
    metadata: { now },
  };

  await bus.send(command);
}

/**
 * Synchronizes weekly planning markdown into task aggregates (Python sync_weekly_plan parity).
 */
export async function syncWeeklyPlan(
  eventStore: MongoDBEventStore,
  bus: EntityCommandBus,
  force = false,
): Promise<SyncWeeklyPlanResult> {
  const settings = await loadSettingsMap(eventStore);

  const syncEnabled = settings[PLANNING_SYNC_ENABLED_KEY];
  if (!force && syncEnabled === "false") {
    return {
      status: "skipped",
      reason:
        "Automatic weekly plan sync is disabled. Use the weekly-planning-assistant CLI (apply-changes / seed-week) or POST with ?force=true.",
      file_parsed: "",
      parser_mode: "skipped",
      tasks_created: 0,
      tasks_updated: 0,
      tasks_archived: 0,
      total_parsed: 0,
    };
  }

  const planningDir = resolvePlanningDir(settings);
  const { filePath, fileName } = getActivePlanningFile(planningDir);

  const fileContent = readFileSync(filePath, "utf-8");
  const fileHash = computeFileHash(fileContent);
  const storedData = parseStoredHash(settings);

  if (
    !force &&
    storedData.file_name === fileName &&
    storedData.hash === fileHash
  ) {
    return {
      status: "skipped",
      reason: "File content hash has not changed.",
      file_parsed: fileName,
      parser_mode: "skipped",
      tasks_created: 0,
      tasks_updated: 0,
      tasks_archived: 0,
      total_parsed: 0,
    };
  }

  let parsedTasks = parseWeeklyPlan(fileContent);
  let parserMode: "regex" | "ai" = "regex";

  if (parsedTasks.length === 0) {
    const apiKey = getKimiApiKey(settings);
    if (apiKey) {
      try {
        const { parseWeeklyPlanAi } = await import("@/ai/syncPlanning/parseWeeklyPlanAi.js");
        const aiResult = await parseWeeklyPlanAi(fileContent, apiKey);
        parsedTasks = aiResult.tasks ?? [];
        if (parsedTasks.length > 0) {
          parserMode = "ai";
        }
      } catch (error) {
        console.warn(
          "AI parsing fallback error:",
          error instanceof Error ? error.message : String(error),
        );
      }
    }
  }

  const projectMap = await loadProjectMap(eventStore);
  const now = new Date();
  let createdCount = 0;
  let updatedCount = 0;

  for (const taskData of parsedTasks) {
    const existingTask = await findTaskByNameAndSource(
      eventStore,
      taskData.name,
      "planning",
    );

    const projectId = guessProjectId(
      taskData.name,
      taskData.description,
      taskData.category,
      projectMap,
    );

    if (existingTask) {
      await handleUpdateTask(bus, existingTask, taskData, projectId, now);
      updatedCount += 1;
    } else {
      await handleCreateTask(bus, taskData, projectId, now);
      createdCount += 1;
    }
  }

  const parsedNames = new Set(parsedTasks.map((t) => t.name));
  const doneTasks = await findDonePlanningTasks(eventStore);
  let archivedCount = 0;

  for (const doneTask of doneTasks) {
    if (!parsedNames.has(doneTask.name)) {
      await handleArchiveTask(bus, doneTask, now);
      archivedCount += 1;
    }
  }

  await upsertSettingValue(
    bus,
    SYNC_HASH_SETTING_KEY,
    JSON.stringify({ file_name: fileName, hash: fileHash }),
    now,
  );

  return {
    status: "success",
    file_parsed: fileName,
    parser_mode: parserMode,
    tasks_created: createdCount,
    tasks_updated: updatedCount,
    tasks_archived: archivedCount,
    total_parsed: parsedTasks.length,
  };
}

File: backend/services/sync_service.py
import os
import re
from datetime import datetime
from typing import List, Dict, Any
from sqlmodel import Session, select
from database import Task, Project, Setting, sync_task_to_record

PLANNING_DIR = r""

def get_active_planning_file() -> str:
    if not os.path.exists(PLANNING_DIR):
        raise FileNotFoundError(f"Planning directory not found at: {PLANNING_DIR}")
    
    files = os.listdir(PLANNING_DIR)
    current_planning_files = [f for f in files if f.startswith("Current_Planning_") and f.endswith(".md")]
    
    if not current_planning_files:
        raise FileNotFoundError("No active weekly planning file (prefixed with 'Current_Planning_') found.")
    
    # Sort just in case there are multiple, grab the latest alphabetically
    current_planning_files.sort()
    return os.path.join(PLANNING_DIR, current_planning_files[-1])

def parse_weekly_plan(file_path: str) -> List[Dict[str, Any]]:
    tasks = []
    
    # Regex to parse checklist line:
    # Matches: - [ ] Ⓑ **Task Name**: description
    # Group 1: status (space, /, x, X)
    # Group 2: owner (Ⓑ, 🅾️, 🤝)
    # Group 3: task title
    # Group 4: description/detail
    task_re = re.compile(r'^\s*-\s*\[([ xX/])\]\s*([Ⓑ🅾️🤝])\s*(?:🔁\s*)?\*\*(.*?)\*\*:?\s*(.*)$')
    
    with open(file_path, "r", encoding="utf-8") as f:
        lines = f.readlines()
        
    in_checklist = False
    current_category = "business" # default
    
    for line in lines:
        line_stripped = line.strip()
        
        # Check if we enter the checklist section
        if line_stripped.startswith("### Week:"):
            in_checklist = True
            continue
            
        # Check if we exit the checklist section
        if in_checklist and (line_stripped.startswith("---") or (line_stripped.startswith("#") and not line_stripped.startswith("####"))):
            in_checklist = False
            
        if not in_checklist:
            continue
            
        # Parse category headers
        if line_stripped.startswith("####"):
            header_text = line_stripped.lower()
            if "development" in header_text or "implementation" in header_text or "💻" in header_text:
                current_category = "dev"
            else:
                current_category = "business"
            continue
            
        # Parse task checkbox
        match = task_re.match(line_stripped)
        if match:
            status_char = match.group(1).lower()
            owner_char = match.group(2)
            title = match.group(3).strip()
            desc = match.group(4).strip()
            
            # Map status
            if status_char == 'x':
                status = "done"
            elif status_char == '/':
                status = "in_progress"
            else:
                status = "backlog"
                
            # Map owner prefix for description
            owner_text = "Shared"
            if owner_char == 'Ⓑ':
                owner_text = "Alice"
            elif owner_char == '🅾️':
                owner_text = "Bob"
                
            full_description = f"Owner: {owner_text} | {desc}" if desc else f"Owner: {owner_text}"
            
            tasks.append({
                "name": title,
                "description": full_description,
                "status": status,
                "category": current_category,
                "owner": owner_text
            })
            
    return tasks

def sync_weekly_plan(session: Session, force: bool = False) -> Dict[str, Any]:
    import hashlib
    import json
    
    file_path = get_active_planning_file()
    file_name = os.path.basename(file_path)
    
    # Read file and compute hash
    with open(file_path, "r", encoding="utf-8") as f:
        file_content = f.read()
        
    file_hash = hashlib.sha256(file_content.encode("utf-8")).hexdigest()
    
    # Fetch stored hash
    hash_setting = session.exec(
        select(Setting).where(Setting.key == "sync_active_file_hash")
    ).first()
    stored_data = json.loads(hash_setting.value) if (hash_setting and hash_setting.value) else {}
    
    if not force and stored_data.get("file_name") == file_name and stored_data.get("hash") == file_hash:
        return {
            "status": "skipped",
            "reason": "File content hash has not changed.",
            "file_parsed": file_name,
            "parser_mode": "skipped",
            "tasks_created": 0,
            "tasks_updated": 0,
            "tasks_archived": 0,
            "total_parsed": 0
        }
        
    parsed_tasks = parse_weekly_plan(file_path)
    
    # AI Fallback: If structured regex parsing yields no tasks, invoke Gemini
    is_ai_parsed = False
    if not parsed_tasks:
        print("Regex parsing returned 0 tasks. Invoking AI parsing fallback...")
        try:
            # Fetch API Key from settings
            api_key_setting = session.exec(
                select(Setting).where((Setting.key == "GEMINI_API_KEY") | (Setting.key == "gemini_api_key"))
            ).first()
            api_key = api_key_setting.value if api_key_setting else None
            
            from services.gemini_service import GeminiService
            gemini = GeminiService(api_key=api_key)
            ai_result = gemini.parse_weekly_plan_ai(file_content)
            parsed_tasks = ai_result.get("tasks", [])
            if parsed_tasks:
                is_ai_parsed = True
                print(f"AI parsing fallback successful. Extracted {len(parsed_tasks)} tasks.")
        except Exception as e:
            print(f"AI parsing fallback error: {e}")
            
    # Get all projects to perform matching
    projects = session.exec(select(Project)).all()
    project_map = {p.name.lower(): p.id for p in projects}
    
    def guess_project_id(title: str, desc: str, category: str) -> str:
        return next(iter(project_map.values()), None)
        
    def guess_task_grouping(title: str, desc: str, category: str) -> str:
        text = (title + " " + desc).lower()
        if category == "dev":
            if "chat" in text or "sse" in text or "pipeline" in text or "middleware" in text:
                return "AI"
            if "access" in text or "auth" in text or "orguser" in text or "invitation" in text:
                return "Access Control"
            if "dsm" in text or "enrich" in text:
                return "Backend"
            if "zod" in text or "validator" in text:
                return "Testing"
            return "AI"
        else:
            if "workflow" in text:
                return "DevOps"
            if "distribution" in text:
                return "Infrastructure"
            if "widget" in text:
                return "Frontend"
            if "demo" in text:
                return "Design"
            if "icp" in text:
                return "Documentation"
            if "linkedin" in text or "comment" in text or "post" in text:
                return "Documentation"
            if "agent" in text or "productivity" in text:
                return "DevOps"
            return "Documentation"
            
    created_count = 0
    updated_count = 0
    
    for task_data in parsed_tasks:
        # Check if task already exists
        existing_task = session.exec(
            select(Task).where(Task.name == task_data["name"], Task.source == "planning")
        ).first()
        
        project_id = guess_project_id(task_data["name"], task_data["description"], task_data["category"])
        owner = task_data.get("owner", "Alice")
        grouping = guess_task_grouping(task_data["name"], task_data["description"], task_data["category"])
        
        if existing_task:
            # Update fields if they changed
            existing_task.description = task_data["description"]
            existing_task.status = task_data["status"]
            existing_task.category = task_data["category"]
            existing_task.owner = owner
            existing_task.task_grouping = grouping
            if project_id:
                existing_task.project_id = project_id
            existing_task.updated_at = datetime.utcnow()
            session.add(existing_task)
            session.commit()
            
            # Sync to EAV DatabaseRecord
            try:
                sync_task_to_record(session, existing_task)
            except Exception as e:
                print(f"Error syncing task to dynamic record: {e}")
                
            updated_count += 1
        else:
            # Create new task
            new_task = Task(
                name=task_data["name"],
                description=task_data["description"],
                status=task_data["status"],
                category=task_data["category"],
                source="planning",
                project_id=project_id,
                owner=owner,
                task_grouping=grouping,
                archived=False
            )
            session.add(new_task)
            session.commit()
            session.refresh(new_task)
            
            # Sync to EAV DatabaseRecord
            try:
                sync_task_to_record(session, new_task)
            except Exception as e:
                print(f"Error syncing task to dynamic record: {e}")
                
            created_count += 1
            
    # Auto-archive done tasks completed in previous weeks
    parsed_names = {t["name"] for t in parsed_tasks}
    done_tasks = session.exec(
        select(Task).where(Task.source == "planning", Task.status == "done", Task.archived == False)
    ).all()
    
    archived_count = 0
    for dt in done_tasks:
        if dt.name not in parsed_names:
            dt.archived = True
            dt.updated_at = datetime.utcnow()
            session.add(dt)
            try:
                sync_task_to_record(session, dt)
            except Exception as e:
                print(f"Error syncing archived task to dynamic record: {e}")
            archived_count += 1
            
    session.commit()
    
    # Save hash settings
    if not hash_setting:
        hash_setting = Setting(key="sync_active_file_hash", value="")
    hash_setting.value = json.dumps({"file_name": file_name, "hash": file_hash})
    session.add(hash_setting)
    session.commit()
    
    return {
        "status": "success",
        "file_parsed": file_name,
        "parser_mode": "ai" if is_ai_parsed else "regex",
        "tasks_created": created_count,
        "tasks_updated": updated_count,
        "tasks_archived": archived_count,
        "total_parsed": len(parsed_tasks)
    }
